Added reference tests to ReductionOpInfo and ported mean test

heitorschueroff opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #63978
* __->__ #62900

fixes https://github.com/pytorch/pytorch/issues/59550

Differential Revision: [D30408815](https://our.internmc.facebook.com/intern/diff/D30408815)

heitorschueroff committed:
Added reference tests to ReductionOpInfo

[ghstack-poisoned]

heitorschueroff committed:
Update on "Added reference tests to ReductionOpInfo"




[ghstack-poisoned]

heitorschueroff committed:
Update on "Added reference tests to ReductionOpInfo"




[ghstack-poisoned]

heitorschueroff committed:
Update on "Added reference tests to ReductionOpInfo"




[ghstack-poisoned]

heitorschueroff committed:
Update on "Added reference tests to ReductionOpInfo"




[ghstack-poisoned]

mruberry commented on test/test_reductions.py:
```diff
@@ -241,6 +243,60 @@ def test_nan_policy_omit(self, device, dtype, op: ReductionOpInfo):
+                    ref = op(t.contiguous(), *args, dim=dim, keepdim=keepdim, **kwargs)
+                    self.assertEqual(res, ref)
+
+    # NumPy does not support BFloat16 so we don't test that agains
```
Nice comment

mruberry commented on test/test_reductions.py:
```diff
@@ -241,6 +243,60 @@ def test_nan_policy_omit(self, device, dtype, op: ReductionOpInfo):
+         allowed_dtypes=get_all_dtypes(include_bfloat16=False))
+    def test_ref_random_input_large(self, device, dtype, op: ReductionOpInfo):
+        """Tests reducing large tensors for accuracy"""
+        t = make_tensor([10000000], device, dtype)
```
power of 2, maybe?

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -5212,6 +5218,18 @@ def gradcheck_wrapper_triangular_input(op, input, *args, upper=False, **kwargs):
+        if t.ndim == 0 and (dim == 0 or dim == -1):
+            dim = None
+
+        return f(t.cpu().numpy(), *args, axis=dim, keepdims=keepdim, **kwargs)
```
.detach().cpu().numpy()?

mruberry reviewed (commented):
Just a few comments/suggestions inline. The thorough expansion of testing is great. Test time is always a looming concern, however. When all the new tests are added and old tests made redundant are removed we should diff the before/after e2e ci time.

heitorschueroff commented on test/test_reductions.py:
```diff
@@ -241,6 +243,60 @@ def test_nan_policy_omit(self, device, dtype, op: ReductionOpInfo):
+    @ops(reduction_op_db)
+    def test_non_contiguous_input(self, device, dtype, op: ReductionOpInfo):
+        """Tests reducing along non contiguous dimensions"""
+        for ndim in range(1, 4):
```
Revised to manually generate some test cases

heitorschueroff commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -906,6 +910,7 @@ def sample_inputs_func(*args, **kwargs):
             return result
 
         # Override OpInfo defaults and call base class __init__
+        kwargs['ref'] = None
```
Updated to use OpInfo ref

heitorschueroff commented on test/test_reductions.py:
```diff
@@ -241,6 +243,60 @@ def test_nan_policy_omit(self, device, dtype, op: ReductionOpInfo):
+    # already specific test for checking output dtype.
+
+    @ops(filter(lambda op: op.reference is not None, reduction_op_db),
+         allowed_dtypes=get_all_dtypes(include_bfloat16=False))
```
The tests are running pretty fast right now, also this is testing for the default test dtypes, if we want to constrain it we should do it through the opinfo test_dtypes kwarg.

heitorschueroff committed:
Update on "Added reference tests to ReductionOpInfo"




[ghstack-poisoned]

heitorschueroff committed:
Update on "Added reference tests to ReductionOpInfo"




[ghstack-poisoned]

heitorschueroff committed:
Update on "Added reference tests to ReductionOpInfo"




[ghstack-poisoned]

heitorschueroff committed:
Update on "Added reference tests to ReductionOpInfo"




[ghstack-poisoned]

mruberry commented on test/test_reductions.py:
```diff
@@ -293,6 +294,96 @@ def test_empty_tensor_nonempty_slice(self, device, op: ReductionOpInfo):
+    @ops(reduction_ops)
+    def test_noncontiguous_innermost(self, device, dtype, op: ReductionOpInfo):
+        """Tests reducing along noncontiguous innermost dimension."""
+        t = make_tensor((10, 10), device, dtype)
```
make_tensor has a noncontiguous kwarg

https://github.com/pytorch/pytorch/blob/fcc1f87b6aacdfba9193691568bc405be42ac77e/torch/testing/_internal/common_utils.py#L1942

mruberry replied:
I guess the issue with using that is you want to control the noncontiguity precisely here

heitorschueroff replied:
Yeah, in this case I'd like to test for different types of non-contiguity

mruberry commented on test/test_reductions.py:
```diff
@@ -293,6 +294,96 @@ def test_empty_tensor_nonempty_slice(self, device, op: ReductionOpInfo):
+    @ops(reduction_ops)
+    def test_noncontiguous_all(self, device, dtype, op: ReductionOpInfo):
+        """Tests reducing all dimensions of a noncontiguous tensor."""
+        t = make_tensor((10, 10), device, dtype)
```
3D might be more interesting for this test

mruberry commented on test/test_reductions.py:
```diff
@@ -293,6 +294,96 @@ def test_empty_tensor_nonempty_slice(self, device, op: ReductionOpInfo):
+    # implementations. We also don't compare dtypes or test for different
+    # keepdim because we already have other tests covering those.
+
+    def _test_ref(self, op: ReductionOpInfo, t: torch.Tensor, **reduction_kwargs):
```
Maybe add a comment about how these reference tests are distinct from

https://github.com/pytorch/pytorch/blob/feba6806c9b7a5e13ffd5839f8bc6ddc2e016a26/test/test_ops.py#L199

which only uses sample inputs

mruberry commented on test/test_reductions.py:
```diff
@@ -293,6 +294,96 @@ def test_empty_tensor_nonempty_slice(self, device, op: ReductionOpInfo):
+
+    @ops(filter(lambda op: op.ref is not None, reduction_ops),
+         allowed_dtypes=get_all_dtypes(include_bfloat16=False))
+    def test_ref_extremal_values(self, device, dtype, op: ReductionOpInfo):
```
Running this on integer types doesn't seem interesting. In the interest of test time float32 and complex64 is probably OK?

mruberry commented on test/test_reductions.py:
```diff
@@ -293,6 +294,96 @@ def test_empty_tensor_nonempty_slice(self, device, op: ReductionOpInfo):
+            self._test_ref(op, make(10, 1, 7, 4), dim=(0, -1))
+            self._test_ref(op, make(10, 1, 7, 4), dim=(0, 1, 3))
+
+    @ops(filter(lambda op: op.ref is not None, reduction_ops),
```
Maybe a subset of dtypes, like uint8, int16, long, float16, float32, and complex64 would alleviate test time?

heitorschueroff replied:
I set the default_test_dtypes to just a subset and by default @ops uses the default_test_dtypes if it is set.

mruberry replied:
Yes but I mean for these reference tests in particular it might be useful to winnow the dtypes down (and not to the default set because it contains bfloat16)

mruberry commented on test/test_reductions.py:
```diff
@@ -293,6 +294,96 @@ def test_empty_tensor_nonempty_slice(self, device, op: ReductionOpInfo):
+            expected = op.ref(t.detach().cpu().numpy(), *args, **kwargs)
+            self.assertEqual(result, expected, exact_dtype=False)
+
+    @ops(filter(lambda op: op.ref is not None, reduction_ops),
```
See note about running on a subset of dtypes below

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -5384,6 +5386,53 @@ def gradcheck_wrapper_triangular_input(op, input, *args, upper=False, **kwargs):
 
 
+def reference_reduction_numpy(f, supports_keepdims=True):
+    """Wraps a NumPy reduction operator.
```
Nice docstring

mruberry reviewed (approved):
Just a few small suggestions inline for your review

heitorschueroff committed:
Update on "Added reference tests to ReductionOpInfo"




[ghstack-poisoned]

heitorschueroff committed:
Update on "Added reference tests to ReductionOpInfo"


Differential Revision: [D30408815](https://our.internmc.facebook.com/intern/diff/D30408815)

[ghstack-poisoned]

heitorschueroff committed:
Update on "Added reference tests to ReductionOpInfo"


Differential Revision: [D30408815](https://our.internmc.facebook.com/intern/diff/D30408815)

[ghstack-poisoned]

heitorschueroff commented:
@heitorschueroff has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30408815).

heitorschueroff committed:
Update on "Added reference tests to ReductionOpInfo"


fixes https://github.com/pytorch/pytorch/issues/59550

Differential Revision: [D30408815](https://our.internmc.facebook.com/intern/diff/D30408815)

[ghstack-poisoned]

heitorschueroff committed:
Update on "Added reference tests to ReductionOpInfo"


fixes https://github.com/pytorch/pytorch/issues/59550

Differential Revision: [D30408815](https://our.internmc.facebook.com/intern/diff/D30408815)

[ghstack-poisoned]

heitorschueroff committed:
Update on "Added reference tests to ReductionOpInfo and ported mean test"


fixes https://github.com/pytorch/pytorch/issues/59550

Differential Revision: [D30408815](https://our.internmc.facebook.com/intern/diff/D30408815)

[ghstack-poisoned]

heitorschueroff commented:
@heitorschueroff has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30408815).

heitorschueroff committed:
Update on "Added reference tests to ReductionOpInfo and ported mean test"


fixes https://github.com/pytorch/pytorch/issues/59550

Differential Revision: [D30408815](https://our.internmc.facebook.com/intern/diff/D30408815)

[ghstack-poisoned]

heitorschueroff commented:
@heitorschueroff has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30408815).

heitorschueroff committed:
Update on "Added reference tests to ReductionOpInfo and ported mean test"


fixes https://github.com/pytorch/pytorch/issues/59550

Differential Revision: [D30408815](https://our.internmc.facebook.com/intern/diff/D30408815)

[ghstack-poisoned]

heitorschueroff commented:
@heitorschueroff has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30408815).

heitorschueroff committed:
Update on "Added reference tests to ReductionOpInfo and ported mean test"


fixes https://github.com/pytorch/pytorch/issues/59550

Differential Revision: [D30408815](https://our.internmc.facebook.com/intern/diff/D30408815)

[ghstack-poisoned]

heitorschueroff committed:
Update on "Added reference tests to ReductionOpInfo and ported mean test"


fixes https://github.com/pytorch/pytorch/issues/59550

Differential Revision: [D30408815](https://our.internmc.facebook.com/intern/diff/D30408815)

[ghstack-poisoned]

heitorschueroff committed:
Update on "Added reference tests to ReductionOpInfo and ported mean test"


fixes https://github.com/pytorch/pytorch/issues/59550

Differential Revision: [D30408815](https://our.internmc.facebook.com/intern/diff/D30408815)

[ghstack-poisoned]
